nixos/test-driver: minor fixes for `nixos-build-vms(8)`

Ma27 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Follow-up for https://github.com/NixOS/nixpkgs/pull/125979. While I agree with the overall approach from said PR, I don't really like the current implications for `nixos-build-vms(8)`. This PR aims to fix that.

Further details for the actual changes can be found in the commit-messages.

cc @blaggacao 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Ma27 committed:
nixos/testing: re-add nixos-run-vms script

Originally removed in 926fb9396881202e727e5ec1fbf609b64455b388. This
one is actually quite useful for `nixos-build-vms(8)`.

Ma27 committed:
nixos/test-driver: start interactive mode if `testScript` is empty

This is relevant for `nixos-build-vms(8)` which doesn't have a
test-script. In that case it's more intuitive to directly go into the
interactive mode which is IMHO more intuitive.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -1061,7 +1062,8 @@ def clean_up() -> None:
                 process.terminate()
         log.close()
 
+    interactive = args.interactive or (not bool(testscript))
```
nice!

blaggacao commented on nixos/lib/testing-python.nix:
```diff
@@ -186,6 +186,12 @@ rec {
           --set startScripts "''${vmStartScripts[*]}" \
           --set testScript "$out/test-script" \
           --set vlans '${toString vlans}'
+
+        ln -s ${testDriver}/bin/nixos-test-driver $out/bin/nixos-run-vms
+        wrapProgram $out/bin/nixos-run-vms \
+          --set startScripts "''${vmStartScripts[*]}" \
+          --set testScript "${pkgs.writeText "start-all" "start_all(); join_all();"}" \
+          --set vlans '${toString vlans}'
```
would it be possible to render this somwhere else, instead? I'd not be totally unhappy, if we could keep this entrypoint here minimal, so that we ca evolve a (single) consistent api, here.

Ma27 replied:
Not sure if I'm misunderstanding you, but since it's a symlink to the test-driver it's (more or less) a wrapper already, correct?

Do you have a better place in mind already? E.g. only adding it to `$out` if `nixos-build-vms` is used rather than an "actual" VM test? Not sure if I like that (for consistency reasons), but I'd like to understand what you have in mind for now :)

blaggacao replied:
> E.g. only adding it to $out if nixos-build-vms is used rather than an "actual" VM test?

Oh, of that was possible, I would indeed favor it.

The reason is that I think we shouldn't broadly advertise an interface through its existence in a place where I have a feeling "less is more".

As the author of this refacor I got a pretty good sense of the situations we can get in if there is a "convolution of interfaces".

So I'm just worried, in principle, that if it's here (instead of there), them over the eons of times undesired ideas might sneak in again and slowly destructure the code into minestrone soup (again).

On the other hand, calling site interfaces ("$out if nixos-build-vms") are a beautiful thing, to avoid this.

If I couldn't convince you, I'm fine, but if I could, I'm happy too. :smile:

Ma27 replied:
Since I use `nixos-build-vms(8)` rather frequently as a maintainer to reproduce bugs or to test modules, I have a fairly strong opinion about it, but the test-driver itself can work without `nixos-run-vms` as well (in other words: no strong opinion about).

I'll check today if there's a reasonably trivial way to implement this and if yes, I'd just make that change.

Ma27 replied:
Updated the PR :) 
Is the removal from the API surface sufficient for you then?

blaggacao replied:
this is an actual improvement! nice! yep, that's completely fine.

tfc commented:
Looks great, let's run some tests to be sure:

@GrahamcOfBorg test bittorrent
@GrahamcOfBorg test boot.biosCdrom
@GrahamcOfBorg test boot.biosNetboot
@GrahamcOfBorg test boot.biosUsb
@GrahamcOfBorg test boot.uefiCdrom
@GrahamcOfBorg test boot.uefiNetboot
@GrahamcOfBorg test boot.uefiUsb
@GrahamcOfBorg test installer.bcache
@GrahamcOfBorg test installer.btrfsSimple
@GrahamcOfBorg test installer.btrfsSubvolDefault
@GrahamcOfBorg test installer.btrfsSubvols
@GrahamcOfBorg test installer.encryptedFSWithKeyfile
@GrahamcOfBorg test installer.grub1
@GrahamcOfBorg test installer.luksroot
@GrahamcOfBorg test installer.lvm
@GrahamcOfBorg test installer.separateBoot
@GrahamcOfBorg test installer.simple
@GrahamcOfBorg test installer.swraid
@GrahamcOfBorg test installer.fsroot

Ma27 committed:
nixos/testing: only create `nixos-run-vms` for `nixos-build-vms(8)`

This is the case when the test-script is empty. `nixos-build-vms(8)` is
primarily supposed to be used as tool to test changes or to reproduce
bugs (IMHO) where "just spinning up a few VMs" is the primary use-case.

In the ongoing discussion about these changes[1] it was suggested to
only expose it when needed (i.e. in the case I described above) to keep
the API surface as slim as possible.

[1] https://github.com/NixOS/nixpkgs/pull/133675#discussion_r688112485

Ma27 commented:
cc @tfc anything to add? :)
